lib/schema-controller.js

```diff
@@ -0,0 +1,169 @@
+'use strict'
+
+const { Schemas } = require('./schemas')
+
+const Ajv = require('ajv')
+const fastJsonStringify = require('fast-json-stringify')
+
+let buildPerformanceValidator
+
+function SchemaController () {
+  this.schemaBucket = new Schemas()
+  this.validatorCompiler = undefined
+  this.serializerCompiler = undefined
+  this.parent = undefined
+}
+
+/**
+ * This method is called once on the fastify server inizialization.
+ * You may use it to set some globals argument as caching
+ */
+SchemaController.prototype.setup = function () {
+  // it is needed once at the server built
+  buildPerformanceValidator = ValidatorSelector()
+}
+
+/**
+ * Called at every fastify context that is being created.
+ * This should return a new SchemaController.
+ * WARN: If it returns `this`, the JSON schema encapsulation will be broken
+ */
+SchemaController.prototype.clone = function clone () {
```

Comment:
Why do we need this method? Again, I'm not clear why we would need schemas to be defined per encapsulation context.

Reply:
This is the (worse) use case testes:

- context A-1 has:
  - schema $id foo
  - context A-2 has:
    - schema $id foo (inherited)
    - schema $id bar
  - context A-3 has:
    - schema $id foo (inherited)
    - schema $id hello
  - context A-4 has:
    - schema $id foo (inherited)
 - context B-1 has:
   - schema $id foo
   - schema $id bar

So in this case there will be created:

- one `ajv` instance per `A-1` with schema `foo`
- one `ajv` instance per `A-2` with schemas `foo, bar`
- one `ajv` instance per `A-3` with schemas `foo, hello`
- the instance `A-4` will inherit the `ajv` instance of `A-1` (per the [JSON.stringify](https://github.com/fastify/fastify/blob/5c7d23fb708b87887360a56db0c5a71458aac4e2/lib/schema-compilers.js#L16) optimization cc @mcollina )
- one `ajv` instance per `A-1` with schema `foo`
- the instance `B-1` will inherit the `ajv` instance of `A-2` (per the [JSON.stringify](https://github.com/fastify/fastify/blob/5c7d23fb708b87887360a56db0c5a71458aac4e2/lib/schema-compilers.js#L16) optimization)

So, for example, `A-3` can't use `bar` and `A-2` can't use `hello` etc..

For this reason, whenever a fastify context is created, the `schemaController` with its schemas is cloned.

This is already in place here:

https://github.com/fastify/fastify/blob/5c7d23fb708b87887360a56db0c5a71458aac4e2/lib/pluginOverride.js#L46

Moving the code into this `schemaController` pointed out some good optimization we could add, working on it.

Reply:
A schema container may only have one schema with `$id: 'foo'`. Therefore, it doesn't make any sense to try and apply encapsulation such that multiple schemas with the same id can be defined. This is per spec -- https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.8.2.2

Basically, `$id` is a URI and a URI by definition points to a singular entity.